File: voice-note-client/android/app/src/main/kotlin/com/spark/suikouji/audio/BargeInDetector.kt
package com.spark.suikouji.audio

import kotlin.math.sqrt

data class BargeInConfig(
    val enabled: Boolean = true,
    val energyThreshold: Double = 0.5,
    val minSpeechMs: Int = 120,
    val cooldownMs: Int = 300,
)

class BargeInDetector(
    private val sampleRate: Int = 16_000,
    private val onTriggered: () -> Unit,
) {
    // Simple energy-based detector with hold-time + cooldown guards.
    @Volatile private var config: BargeInConfig = BargeInConfig()
    @Volatile private var speechMs: Int = 0
    @Volatile private var lastTriggerTs: Long = 0L

    fun updateConfig(newConfig: BargeInConfig) {
        config = newConfig
        speechMs = 0
    }

    fun onFrame(frame: ByteArray, ttsPlaying: Boolean) {
        val cfg = config
        if (!cfg.enabled || !ttsPlaying) {
            speechMs = 0
            return
        }

        val now = System.currentTimeMillis()
        if (now - lastTriggerTs < cfg.cooldownMs) return

        val rms = normalizedRms(frame)
        if (rms >= cfg.energyThreshold) {
            // Require continuous speech window to reduce accidental triggers.
            val frameMs = (frame.size / 2.0 / sampleRate * 1000).toInt().coerceAtLeast(1)
            speechMs += frameMs
            if (speechMs >= cfg.minSpeechMs) {
                lastTriggerTs = now
                speechMs = 0
                onTriggered()
            }
        } else {
            speechMs = 0
        }
    }

    private fun normalizedRms(frame: ByteArray): Double {
        if (frame.size < 2) return 0.0
        var sum = 0.0
        var count = 0
        var i = 0
        while (i + 1 < frame.size) {
            val sample = ((frame[i + 1].toInt() shl 8) or (frame[i].toInt() and 0xFF)).toShort()
            val n = sample.toDouble() / Short.MAX_VALUE.toDouble()
            sum += n * n
            count++
            i += 2
        }
        if (count == 0) return 0.0
        return sqrt(sum / count)
    }
}

File: voice-note-client/ios/Runner/Audio/BargeInDetector.swift
import Foundation

struct BargeInConfig {
  let enabled: Bool
  let energyThreshold: Double
  let minSpeechMs: Int
  let cooldownMs: Int

  static let `default` = BargeInConfig(
    enabled: true,
    energyThreshold: 0.5,
    minSpeechMs: 120,
    cooldownMs: 300
  )
}

final class BargeInDetector {
  private let sampleRate: Double
  private let onTriggered: () -> Void

  // Energy VAD with hold-time and cooldown to reduce false positives.
  private var config: BargeInConfig = .default
  private var speechMs = 0
  private var lastTriggerAtMs = 0

  init(sampleRate: Double = 16_000, onTriggered: @escaping () -> Void) {
    self.sampleRate = sampleRate
    self.onTriggered = onTriggered
  }

  func updateConfig(_ config: BargeInConfig) {
    self.config = config
    speechMs = 0
  }

  func onFrame(_ frame: Data, ttsPlaying: Bool) {
    guard config.enabled, ttsPlaying else {
      speechMs = 0
      return
    }

    let now = Int(Date().timeIntervalSince1970 * 1000)
    guard now - lastTriggerAtMs >= config.cooldownMs else { return }

    let rms = normalizedRms(frame)
    if rms >= config.energyThreshold {
      let frameSampleCount = max(1, frame.count / MemoryLayout<Float>.size)
      let frameDurationMs = max(1, Int((Double(frameSampleCount) / sampleRate) * 1000))
      speechMs += frameDurationMs
      if speechMs >= config.minSpeechMs {
        speechMs = 0
        lastTriggerAtMs = now
        onTriggered()
      }
    } else {
      speechMs = 0
    }
  }

  private func normalizedRms(_ frame: Data) -> Double {
    if frame.count < MemoryLayout<Float>.size { return 0 }
    let sampleCount = frame.count / MemoryLayout<Float>.size
    return frame.withUnsafeBytes { rawBuffer in
      guard let floatPtr = rawBuffer.bindMemory(to: Float.self).baseAddress else {
        return 0
      }
      var sum: Double = 0
      for i in 0..<sampleCount {
        let v = Double(floatPtr[i])
        sum += v * v
      }
      if sampleCount == 0 { return 0 }
      return sqrt(sum / Double(sampleCount))
    }
  }
}
